File: Leetcode/K-th Smallest Prime Fraction/K-th Smallest Prime Fraction.py
#Problem:https://leetcode.com/problems/k-th-smallest-prime-fraction/
class Solution(object):
    def kthSmallestPrimeFraction(self, A, K):
        def count_smaller_fractions(x):     # return number for fractions < x and largest such fraction

            count, denominator, largest = 0, 1, [0, 1]

            for numerator in range(len(A) - 1):     # for each numerator, find first denominator so fraction < x

                while denominator < len(A) and A[numerator] >= x * A[denominator]:  # fraction >= x
                    denominator += 1

                if denominator != len(A) and A[numerator] * largest[1] > largest[0] * A[denominator]:
                    largest = [A[numerator], A[denominator]]                        # new largest

                count += len(A) - denominator       # count this and all greater denominators

            return count, largest

        low, high = 0, 1.0
        while high - low > 10 ** -9:

            mid = (low + high) / 2
            count, largest = count_smaller_fractions(mid)
            if count < K:           # insufficient fractions below mid, search larger candidates
                low = mid
            else:                   # sufficient fractions below mid, search smaller candidates
                result = largest    # update largest fraction where count == K
                high = mid

        return result

File: Leetcode/K-th Smallest Prime Fraction/K-th Smallest Prime Fraction.java

//Java program for kth smallest prime number
class Solution {
    public int[] kthSmallestPrimeFraction(int[] arr, int k) {
        //size of array n
        int n = arr.length;
    double lo = 0.0, hi = 1.0;
    int p = 0, q = 0;
    while (true) {
        double mid = (lo + hi) / 2;
        // find the prime fraction having value less than eqaul to mid
        int count = 0;
        double maxFrac = 0.0;
        int j = 1;
        for (int i = 0; i < n - 1; i++) {
            while (j < n && arr[i] > (arr[j] * mid)) {
                j++;
            }
            //count is added to n-j
            count += (n - j);
            if (j == n)
                break;
            double currFrac = ((double)arr[i] / (double)arr[j]);
            //if currFrac > maxFrac then maxFrac = currFrac
            if (currFrac > maxFrac) {
                p = i;
                q = j;
                maxFrac = currFrac;
            }
        }
        //if count is equal to k then return new int array
        if (count == k) {
            return new int[] { arr[p], arr[q] };
        } else if (count > k) {
            //return hi as mid
            hi = mid;
        } else {
            //return lo as mid
            lo = mid;
        }
    }


    }
}

